py/objdict.c

```diff
@@ -321,8 +321,16 @@ STATIC mp_obj_t dict_popitem(mp_obj_t self_in) {
     mp_check_self(mp_obj_is_dict_type(self_in));
     mp_obj_dict_t *self = MP_OBJ_TO_PTR(self_in);
     mp_ensure_not_fixed(self);
-    size_t cur = 0;
-    mp_map_elem_t *next = dict_iter_next(self, &cur);
+    mp_map_elem_t *next;
+    #if MICROPY_PY_COLLECTIONS_ORDEREDDICT
+    if (self->map.is_ordered) {
+        next = &self->map.table[self->map.used - 1];
```

Comment:
I see there's a test for popitem() when empty, but how does it actually work, wont this just make next point to an invalid entry, rather than making next NULL?

Reply:
Fixed, and slightly optimised it for code size. Now +20 bytes on PYBV11.

I'm a little nervous about setting the key to MP_OBJ_SENTINEL (it should probably NULL to be consistent with `mp_map_lookup`, but the tests pass and I can't find a code path that cares -- i.e. `mp_set_slot_is_filled` handles both). Despite the comment in 9275c182708ae80ac62eaf0e4152df7b88163076 though I don't see a crash while running the tests if I change mp_map_lookup to set it to SENTINEL instead).

Added some asserts and comments for things that I noticed while tracing this.

Reply:
> I'm a little nervous about setting the key to MP_OBJ_SENTINEL

I think the reason it's ok is because the removed object is at the very end, and because `map.used` is decreased the search in `mp_map_lookup()` (which would have failed had it hit a SENTINEL) only searches used entries.